k2/csrc/fsa.cc

```diff
@@ -0,0 +1,250 @@
+// k2/csrc/fsa.cc
+
+// Copyright (c)  2020  Daniel Povey
+//                      Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa.h"
+
+namespace {
+
+// 64-byte alignment should be enough for AVX512 and other computations.
+constexpr size_t kAlignment = 64;
+static_assert((kAlignment & 15) == 0,
+              "kAlignment should be at least multiple of 16");
+static_assert(kAlignment % alignof(k2::Arc) == 0, "");
+static_assert(kAlignment % alignof(k2::CfsaVecHeader) == 0, "");
+
+inline size_t AlignTo(size_t b, size_t alignment) {
+  // alignment should be power of 2
+  return (b + alignment - 1) & (~(alignment - 1));
+}
+
+}  // namespace
+
+namespace k2 {
+
+std::ostream &operator<<(std::ostream &os, const Arc &arc) {
+  os << arc.src_state << " " << arc.dest_state << " " << arc.label;
+  return os;
+}
+
+std::ostream &operator<<(std::ostream &os, const Cfsa &cfsa) {
+  os << "num_states: " << cfsa.num_states << "\n";
+  os << "begin_arc: " << cfsa.begin_arc << "\n";
+  os << "end_arc: " << cfsa.end_arc << "\n";
+  os << "num_arcs: " << cfsa.NumArcs() << "\n";
+  for (int i = cfsa.begin_arc; i != cfsa.end_arc; ++i) {
+    os << cfsa.arcs[i] << "\n";
+  }
+  return os;
+}
+
+Cfsa::Cfsa()
+    : num_states(0),
+      begin_arc(0),
+      end_arc(0),
+      arc_indexes(nullptr),
+      arcs(nullptr) {}
+
+Cfsa::Cfsa(const Fsa &fsa) {
+  begin_arc = 0;
+  num_states = fsa.NumStates();
+  if (num_states != 0) {
+    // this is not an empty fsa
+    arc_indexes = fsa.arc_indexes.data();
+    end_arc = fsa.arc_indexes.back();
+    arcs = const_cast<Arc *>(fsa.arcs.data());
+  } else {
+    // this is an empty fsa
+    arc_indexes = nullptr;
+    end_arc = 0;
+    arcs = nullptr;
+  }
+}
+
+CfsaVec::CfsaVec(size_t size, void *data)
+    : data_(reinterpret_cast<int32_t *>(data)), size_(size) {
+  const auto header = reinterpret_cast<const CfsaVecHeader *>(data_);
+  num_fsas_ = header->num_fsas;
+}
+
+Cfsa CfsaVec::operator[](int32_t f) const {
+  DCHECK_GE(f, 0);
+  DCHECK_LT(f, num_fsas_);
+
+  Cfsa cfsa;
+
+  const auto header = reinterpret_cast<const CfsaVecHeader *>(data_);
```

Comment:
I guess it would be efficient if we cache `header` (and `arcs_array`) in the constructor as we actually do this in `operator[](f)` again and again if user access cfsa in `CfsaVec` one by one, even if `reinterpret_cast` has no runtime cost for pointers (but we excute those statements again and again)

Reply:
I guess the optimizer will optimize it out.

Change it to a data field will occupy extra memory (sizeof a pointer).